Regression: cannot order by boolean in 2.3

We had a Cypher 2.1 query that included a clause like `ORDER BY n.bool DESC` where `n.bool` is a Boolean property.  After upgrading to 2.3, the query throws an exception.  I didn't test 2.2.

```
Caused by: org.neo4j.cypher.IncomparableValuesException: Don't know how to compare that. Left: false (Boolean); Right: false (Boolean)
    at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_3$.incomparableValuesException(CompatibilityFor2_3.scala:67) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_3$.incomparableValuesException(CompatibilityFor2_3.scala:58) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.frontend.v2_3.IncomparableValuesException.mapToPublic(CypherException.scala:115) ~[org.neo4j.neo4j-cypher-frontend-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_3$.runSafely(CompatibilityFor2_3.scala:119) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper.run(CompatibilityFor2_3.scala:191) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.PreparedPlanExecution.execute(PreparedPlanExecution.scala:27) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:118) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:104) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.javacompat.ExecutionEngine.execute(ExecutionEngine.java:87) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.springframework.data.neo4j.support.query.CypherQueryEngineImpl.parseAndExecuteQuery(CypherQueryEngineImpl.java:71) ~[org.springframework.data.spring-data-neo4j-3.4.0.RELEASE.jar:na]
    ... 41 common frames omitted
Caused by: org.neo4j.cypher.internal.frontend.v2_3.IncomparableValuesException: null
    at org.neo4j.cypher.internal.compiler.v2_3.Comparer$class.compare(Comparer.scala:38) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.TopPipe.compare(TopPipe.scala:35) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.TopPipe$LessThanComparator.compare(TopPipe.scala:50) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.TopPipe$LessThanComparator.compare(TopPipe.scala:44) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at java.util.TimSort.countRunAndMakeAscending(TimSort.java:324) ~[na:1.7.0_79]
    at java.util.TimSort.sort(TimSort.java:189) ~[na:1.7.0_79]
    at java.util.TimSort.sort(TimSort.java:173) ~[na:1.7.0_79]
    at java.util.Arrays.sort(Arrays.java:659) ~[na:1.7.0_79]
    at scala.collection.SeqLike$class.sorted(SeqLike.scala:648) ~[org.scala-lang.scala-library-2.11.7.jar:na]
    at scala.collection.mutable.ArrayOps$ofRef.sorted(ArrayOps.scala:186) ~[org.scala-lang.scala-library-2.11.7.jar:na]
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.TopPipe.internalCreateResults(TopPipe.scala:96) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:125) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:122) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:122) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:122) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.createResults(DefaultExecutionResultBuilderFactory.scala:93) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.build(DefaultExecutionResultBuilderFactory.scala:63) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:230) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:214) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anon$3.run(ExecutionPlanBuilder.scala:172) ~[org.neo4j.neo4j-cypher-compiler-2.3-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:192) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:192) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_3$.runSafely(CompatibilityFor2_3.scala:114) ~[org.neo4j.neo4j-cypher-2.3.0.jar:2.3.0]
    ... 47 common frames omitted
```

Workaround:

`ORDER BY (CASE my.bool WHEN TRUE THEN 0 ELSE 1 END)`

--- a/community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/OrderByAcceptanceTest.scala
+++ b/community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/OrderByAcceptanceTest.scala
@@ -19,8 +19,8 @@
  */
 package org.neo4j.internal.cypher.acceptance
 
-import org.neo4j.cypher.{SyntaxException, NewPlannerTestSupport, ExecutionEngineFunSuite}
 import org.neo4j.cypher.internal.compiler.v2_3.test_helpers.CustomMatchers
+import org.neo4j.cypher.{ExecutionEngineFunSuite, NewPlannerTestSupport, SyntaxException}
 
 class OrderByAcceptanceTest extends ExecutionEngineFunSuite with CustomMatchers with NewPlannerTestSupport {
 
@@ -93,5 +93,37 @@ class OrderByAcceptanceTest extends ExecutionEngineFunSuite with CustomMatchers 
     executeWithAllPlanners("MATCH (p:Person) RETURN p ORDER BY p.name LIMIT 0").length should equal(0)
     executeWithAllPlanners("MATCH (p:Person) RETURN p ORDER BY p.name LIMIT {limit}", "limit" -> -1).length should equal(0)
     a [SyntaxException] should be thrownBy executeWithAllPlanners("MATCH (p:Person) RETURN p ORDER BY p.name LIMIT -1")
+  }
+
+  test("should be able to order booleans") {
+    val query = "UNWIND [true, false] AS bools RETURN bools ORDER BY bools"
+
+    val expected = List(Map("bools" -> false), Map("bools" -> true))
+    executeWithAllPlanners(query).toList should equal(expected)
+    executeWithAllPlanners(s"$query DESC").toList should equal(expected.reverse)
+  }
+
+  test("should be able to order strings") {
+    val query = "UNWIND ['.*', '', ' ', 'one'] AS strings RETURN strings ORDER BY strings"
+
+    val expected = List(Map("strings" -> ""), Map("strings" -> " "), Map("strings" -> ".*"), Map("strings" -> "one"))
+    executeWithAllPlanners(query).toList should equal(expected)
+    executeWithAllPlanners(s"$query DESC").toList should equal(expected.reverse)
+  }
+
+  test("should be able to order ints") {
+    val query = "UNWIND [1,3,2] as ints return ints order by ints"
+
+    val expected = List(Map("ints" -> 1), Map("ints" -> 2), Map("ints" -> 3))
+    executeWithAllPlanners(query).toList should equal(expected)
+    executeWithAllPlanners(s"$query DESC").toList should equal(expected.reverse)
+  }
+
+  test("should be able to order floats") {
+    val query = "UNWIND [1.5,1.3,999.99] as floats return floats order by floats"
+
+    val expected = List(Map("floats" -> 1.3), Map("floats" -> 1.5), Map("floats" -> 999.99))
+    executeWithAllPlanners(query).toList should equal(expected)
+    executeWithAllPlanners(s"$query DESC").toList should equal(expected.reverse)
   }
 }

--- a/community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/Comparer.scala
+++ b/community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/Comparer.scala
@@ -32,7 +32,7 @@ trait Comparer extends StringHelper {
 
   def compare(l: Any, r: Any)(implicit qtx: QueryState): Int = {
     try {
-      if ( (isString(l) && isString(r)) || (isNumber(l) && isNumber(r)))
+      if ((isString(l) && isString(r)) || (isNumber(l) && isNumber(r)) || (isBoolean(l) && isBoolean(r)))
         CypherOrdering.DEFAULT.compare(l, r)
       else
         throw new IncomparableValuesException(textWithType(l), textWithType(r))
@@ -52,6 +52,11 @@ object Comparer {
 
   def isNumber(value: Any): Boolean = value match {
     case _: Number => true
+    case _ => value == null
+  }
+
+  def isBoolean(value: Any): Boolean = value match {
+    case _: Boolean => true
     case _ => value == null
   }
 }

